### PR TITLE
Fixes #634 Selecting language for voice input

### LINKED ISSUE #634: Selecting language for voice input
We have some design choices around voice input for multiple languages, the initial backend implementation of which is landing in: #633 

The question is: can the user change the language? I could imagine a few choices:
1) Only allow voice input/search against the current OS language used by the device
2) Change the keyboard/microphone input UX/UI to allow selecting the language (as one might select a different language keyboard for input)
3) Provide some affordance in settings for changing the overall UX/UI language for Firefox Reality

Opinons/feedback? The first option seems easiest, but I assume this (especially the third) might also intersect with other work we are doing.

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #634 #767 Added a new setting to support the new batch of languages. Also rearranged the settings panel.

### CODE REVIEW

File: app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java [hunk 1]
```diff
@@ -305,4 +306,15 @@ public void setMSAALevel(int level) {
         editor.commit();
     }
 
+    public String getVoiceSearchLanguage() {
+        return mPrefs.getString(
+                mContext.getString(R.string.settings_key_voice_search_language), LocaleUtils.getCurrentLocale());
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java]: I don't see `voice_search_language` in `strings.xml` in here or in our [`strings.xml`](https://github.com/MozillaReality/FirefoxReality/blob/096424app/src/main/res/values/strings.xml
) in our PR #711 (but I do see it in the [Wiki page](https://github.com/MozillaReality/FirefoxReality/wiki/L10n#voice_search_language) - thanks!)

Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java]: As per our conversation in Slack we will need to keep those language related strings here for the moment.

Regarding **settings_key_voice_search_language** that's a settings key so not being a translatable string it lives in **non_L10n.xml**.

Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java]: @cvan can you also provide a longer logcat, I don't seem to be able to reproduce that crash in local. I think the real crash is a bit above the one that you sent.

**Comment by keianhzo** [comment]
@cvan Updated the buttons order and paddings.

File: app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java (hunk 1 above)
Comment (cvan) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java]: happened again on another Oculus Go headset:

https://gist.github.com/cvan/675cff1162dc9f5a4cf8153ab3627be8

not sure what to sift through from that

**Comment by keianhzo** [comment]
@cvan The crash should be fixed.

File: app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java (hunk 1 above)
Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/SettingsStore.java]: @cvan The crash should be fixed.

**Comment by keianhzo** [comment]
@daoshengmu could you do a Chinese review?

File: app/src/main/res/layout/settings.xml [hunk 2]
```diff
@@ -68,71 +73,79 @@
...
-                        <Space
-                            android:id="@+id/center"
-                            android:layout_width="0dp"
-                            android:layout_height="0dp"
-                            android:layout_centerHorizontal="true"
-                            android:visibility="invisible" />
+                        <org.mozilla.vrbrowser.ui.views.HoneycombButton
+                            android:id="@+id/displayButton"
+                            style="?attr/settingsButtonStyle"
+                            android:layout_marginRight="10dp"
```
Comment (cvan) [inline app/src/main/res/layout/settings.xml]: can you adjust the vertical centreing of this one?

image

it'd be nice if all the icons are lined up on the bottom row of honeycombs too. thanks in advance!

**Comment by daoshengmu** [comment]
@keianhzo it seems like something wrong, it can't recognize what I speak for Traditional and Simplified Chinese. Moreover, I think we don't need to put `Zhuyin` and `Pinyin` in the option because they are typing inputs not languages. We should consider only show `Traditional Chinese`, `Simplified Chinese`. Render native text is also a good suggestion, it would be `繁體中文` `简体中文`.

File: app/src/main/res/layout/settings.xml (hunk 2 above)
Comment (keianhzo) [reply app/src/main/res/layout/settings.xml]: Fixed buttons alignments.

**Comment by keianhzo** [comment]
@cvan I have fixed most of the requested changes:
- Button alignments
- Languages in their native texts
- I have also addressed the issue #767

@daoshengmu I have fixed the Zhuyin/Pinyin issue. Can you try again see if works. I've tried with probably the only thing I know and seemed to work but of course that's not a great test.

@andrenatal Can you take a look at what @daoshengmu mentions about chinese recognition? I checked that the language for traditional and simplified chinese are the correct ones but just in case.

**Comment by daoshengmu** [comment]
It will crash then I click the mic icon or trying to change the language. I don't have time to trace it. I will take a look at why it crashes on Monday.

**Comment by keianhzo** [comment]
@daoshengmu Fixed

**Comment by andrenatal** [comment]
Hey guys, sorry I'm on PTO until next Monday and don't have my go with me. I'll check this patch when I get back right after thanksgiving. Thanks!

**Comment by cvan** [comment]
@daoshengmu thanks for testing this. that URL issue is significant and a fix (PR #765) has merged into master.

@andrenatal: no worries! we are good on this. will report via email for when you are back. thanks for all your help. Happy Thanksgiving! 🦃

**Comment by cvan** [comment]
@keianhzo:

I changed my language to German and did some testing. nice work!

a few questions before merging:

- can you share some step-by-step tips of how to test this?
- which locales have you tested with / which do you need help with?
- how can I tell that the voice commands are correct or incorrect?

**Comment by keianhzo** [comment]
@cvan Fixed the merge conflicts, feel free to review and merge and we can address the UI fixes in a follow up so everybody can test this ASAP.

Regarding your questions:

> can you share some step-by-step tips of how to test this?

What I do is:
- Select a new language
- Open voice search and say a recognizable word
- Wait until the search finishes and the word should show up in the URL bar

> which locales have you tested with / which do you need help with?
I've done superficial testing for this languages:
- Spanish
- English
- French
- German
- Chinese
- Italian

Superficial means that I just tested common words, just to make sure that is actually identifying the new language. This will require native speaker testing that's also a good reason for landing it if there is anything outstanding.

> how can I tell that the voice commands are correct or incorrect?
Just check if the world that shows up in the Google search bar is correct.

**Comment by cvan** [comment]
My locale is `es-ES` (Spanish). I cannot perform Voice Searches.

* Video: https://vimeo.com/user41954276/review/302177210/ea5eaccb50
* Crash reports: https://gist.github.com/cvan/2c1944a4a3494a4e94101c33491c3a63
* STR:
    1. on fresh install, I click on the Voice Search mic, I accept Oculus permissions, the Voice Search prompt does not show up.
    2. I click on the Voice Search mic icon again.
    3. Stuck at the `!Lo Siento!` screen (`Sorry! I could not understand.`) indefinitely. I waited a good 1 minute+.
    4. Click on the Voice Search mic icon again.
    5. This time, the screen shown is `¿Qué le gustaría buscar en Internet?` (`What would you like to search on the Web?`).
    6. No voice search will work.

**Comment by keianhzo** [comment]
@cvan I've fixed the merge conflicts and the prompt fix has already landed. Can you test again see if you can still reproduce the issue. In my tests it's working fine. Anyway this will require a broader testing by as many people as possible and locales as possible.

**Comment by larsbergstrom** [comment]
I'd like to echo that if we could get this landed soon, it'd be ideal - we really need to be testing this, as I need to submit a 1.1 build on Monday, so having as much of Friday to test widely with a TC-produced build would be super helpful.

**Pull request merged**